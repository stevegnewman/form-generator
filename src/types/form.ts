export interface FormField {
  name: string;
  label: string;
  type: 'text' | 'dropdown';
  options?: string[];
}

export interface FormQuestion {
  title: string;
  fields: FormField[];
}

export interface FormConfig {
  questions: FormQuestion[];
}

export interface FormData {
  [key: string]: string;
}
