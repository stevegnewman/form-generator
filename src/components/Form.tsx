import React from 'react';
import { FormConfig, FormData } from '../types/form';
import { FormField } from './FormField';
import { FormSubmit } from './FormSubmit';

interface FormProps {
  config: FormConfig;
  onSubmit: (data: FormData) => void;
}

/**
 * Dynamic form component that renders based on provided configuration.
 * Handles form state, validation, and submission logic.
 */
export const Form: React.FC<FormProps> = ({ config, onSubmit }) => {
  // Form state management
  const [formData, setFormData] = React.useState<FormData>({});
  const [showSuccess, setShowSuccess] = React.useState(false);
  const [showError, setShowError] = React.useState(false);

  // Handle individual field changes
  const handleFieldChange = (name: string, value: string) => {
    setFormData((prev) => ({
      ...prev,
      [name]: value
    }));
    // Clear error state when user starts typing
    if (showError) {
      setShowError(false);
    }
  };

  // Reset form to initial state
  const clearForm = () => {
    setFormData({});
  };

  // Validate if at least one field is filled
  const isFormValid = () => {
    const fields = config.questions.flatMap(q => q.fields.map(f => f.name));
    return fields.some(fieldName => formData[fieldName] && formData[fieldName].trim() !== '');
  };

  // Handle form submission with validation
  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    
    if (!isFormValid()) {
      setShowError(true);
      setTimeout(() => {
        setShowError(false);
      }, 3000);
      return;
    }

    onSubmit(formData);
    setShowSuccess(true);
    clearForm();

    setTimeout(() => {
      setShowSuccess(false);
    }, 5000);
  };

  return (
    <form onSubmit={handleSubmit}>
      {/* Render form sections based on config */}
      {config.questions.map((question, index) => (
        <div key={index} className="mb-8">
          <h2 className="text-2xl font-semibold mb-4 text-bodyText">{question.title}</h2>
          {/* Render form fields for each section */}
          {question.fields.map((field) => (
            <FormField
              key={field.name}
              field={field}
              value={formData[field.name] || ''}
              onChange={handleFieldChange}
            />
          ))}
        </div>
      ))}
      {/* Submit button with success/error states */}
      <FormSubmit 
        type="submit" 
        fullWidth 
        showSuccess={showSuccess}
        showError={showError}
      >
        Submit
      </FormSubmit>
    </form>
  );
}; 