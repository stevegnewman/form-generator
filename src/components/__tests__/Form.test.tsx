import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import { Form } from '../Form';
import { FormConfig } from '../../types/form';

describe('Form', () => {
  const mockConfig: FormConfig = {
    questions: [
      {
        title: "Test Question",
        fields: [
          { name: "test_field", label: "Test Field", type: "text" },
          { name: "test_dropdown", label: "Test Dropdown", type: "dropdown", options: ["Option 1", "Option 2"] }
        ]
      }
    ]
  };

  const mockOnSubmit = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('renders all fields from config', () => {
    render(<Form config={mockConfig} onSubmit={mockOnSubmit} />);
    
    expect(screen.getByText('Test Question')).toBeInTheDocument();
    expect(screen.getByLabelText('Test Field')).toBeInTheDocument();
    expect(screen.getByLabelText('Test Dropdown')).toBeInTheDocument();
  });

  it('shows error when submitting empty form', async () => {
    render(<Form config={mockConfig} onSubmit={mockOnSubmit} />);
    
    fireEvent.submit(screen.getByRole('button', { name: /submit/i }));
    
    expect(screen.getByText(/please complete form/i)).toBeInTheDocument();
    expect(mockOnSubmit).not.toHaveBeenCalled();

    // Test error message disappears after timeout
    act(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(screen.queryByText(/please complete form/i)).not.toBeInTheDocument();
  });

  it('submits form data when all fields are filled', async () => {
    render(<Form config={mockConfig} onSubmit={mockOnSubmit} />);
    
    fireEvent.change(screen.getByLabelText('Test Field'), {
      target: { value: 'test value' }
    });

    fireEvent.change(screen.getByLabelText('Test Dropdown'), {
      target: { value: 'Option 1' }
    });

    fireEvent.submit(screen.getByRole('button', { name: /submit/i }));

    expect(mockOnSubmit).toHaveBeenCalledWith({
      test_field: 'test value',
      test_dropdown: 'Option 1'
    });
    
    expect(screen.getByText(/form submitted successfully/i)).toBeInTheDocument();

    // Test success message disappears after timeout
    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(screen.queryByText(/form submitted successfully/i)).not.toBeInTheDocument();
  });

  it('clears form after successful submission', async () => {
    render(<Form config={mockConfig} onSubmit={mockOnSubmit} />);
    
    fireEvent.change(screen.getByLabelText('Test Field'), {
      target: { value: 'test value' }
    });

    fireEvent.change(screen.getByLabelText('Test Dropdown'), {
      target: { value: 'Option 1' }
    });

    fireEvent.submit(screen.getByRole('button', { name: /submit/i }));

    expect(screen.getByLabelText('Test Field')).toHaveValue('');
    expect(screen.getByLabelText('Test Dropdown')).toHaveValue('');
  });

  it('clears error state when user starts typing', async () => {
    render(<Form config={mockConfig} onSubmit={mockOnSubmit} />);
    
    // Submit empty form to trigger error
    fireEvent.submit(screen.getByRole('button', { name: /submit/i }));
    expect(screen.getByText(/please complete form/i)).toBeInTheDocument();

    // Start typing in a field
    fireEvent.change(screen.getByLabelText('Test Field'), {
      target: { value: 'a' }
    });

    // Error message should be gone
    expect(screen.queryByText(/please complete form/i)).not.toBeInTheDocument();
  });

  it('validates at least one field is filled before submission', () => {
    render(<Form config={mockConfig} onSubmit={mockOnSubmit} />);
    
    // Submit empty form
    fireEvent.submit(screen.getByRole('button', { name: /submit/i }));
    expect(screen.getByText(/please complete form/i)).toBeInTheDocument();
    expect(mockOnSubmit).not.toHaveBeenCalled();

    // Fill one field and submit
    fireEvent.change(screen.getByLabelText('Test Field'), {
      target: { value: 'test value' }
    });

    fireEvent.submit(screen.getByRole('button', { name: /submit/i }));
    expect(mockOnSubmit).toHaveBeenCalledWith({
      test_field: 'test value'
    });
  });

  it('considers whitespace-only values as empty', () => {
    render(<Form config={mockConfig} onSubmit={mockOnSubmit} />);
    
    // Fill field with whitespace
    fireEvent.change(screen.getByLabelText('Test Field'), {
      target: { value: '   ' }
    });

    fireEvent.submit(screen.getByRole('button', { name: /submit/i }));
    expect(screen.getByText(/please complete form/i)).toBeInTheDocument();
    expect(mockOnSubmit).not.toHaveBeenCalled();
  });

  /* Snapshot tests */
  it('matches snapshot for initial render', () => {
    const { container } = render(
      <Form config={mockConfig} onSubmit={mockOnSubmit} />
    );
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with filled form data', () => {
    const { container } = render(
      <Form config={mockConfig} onSubmit={mockOnSubmit} />
    );

    fireEvent.change(screen.getByLabelText('Test Field'), {
      target: { value: 'test value' }
    });

    fireEvent.change(screen.getByLabelText('Test Dropdown'), {
      target: { value: 'Option 1' }
    });

    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with error state', () => {
    const { container } = render(
      <Form config={mockConfig} onSubmit={mockOnSubmit} />
    );

    fireEvent.submit(screen.getByRole('button'));
    expect(container).toMatchSnapshot();
  });
}); 