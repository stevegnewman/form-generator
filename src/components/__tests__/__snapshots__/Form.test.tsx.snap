// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches snapshot for initial render 1`] = `
<div>
  <form>
    <div
      class="mb-8"
    >
      <h2
        class="text-2xl font-semibold mb-4 text-bodyText"
      >
        Test Question
      </h2>
      <div
        class="mb-4"
      >
        <label
          class="block text-bodyText text-sm font-bold mb-2"
          for="test_field"
        >
          Test Field
        </label>
        <input
          class="appearance-none border rounded w-full py-2 px-3 text-bodyText focus:ring-2 focus:ring-primary focus:border-primary outline-none transition-all duration-200"
          id="test_field"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-4"
      >
        <label
          class="block text-bodyText text-sm font-bold mb-2"
          for="test_dropdown"
        >
          Test Dropdown
        </label>
        <select
          class="border rounded w-full py-2 px-3 text-bodyText focus:ring-2 focus:ring-primary focus:border-primary outline-none transition-all duration-200 cursor-pointer"
          id="test_dropdown"
        >
          <option
            value=""
          >
            Select 
            Test Dropdown
          </option>
          <option
            value="Option 1"
          >
            Option 1
          </option>
          <option
            value="Option 2"
          >
            Option 2
          </option>
        </select>
      </div>
    </div>
    <button
      class="py-[7px] px-[30px] rounded-3xl transition-all duration-200 font-bold bg-primary text-white hover:bg-white hover:text-bodyText border-2 border-primary w-full  "
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form matches snapshot with error state 1`] = `
<div>
  <form>
    <div
      class="mb-8"
    >
      <h2
        class="text-2xl font-semibold mb-4 text-bodyText"
      >
        Test Question
      </h2>
      <div
        class="mb-4"
      >
        <label
          class="block text-bodyText text-sm font-bold mb-2"
          for="test_field"
        >
          Test Field
        </label>
        <input
          class="appearance-none border rounded w-full py-2 px-3 text-bodyText focus:ring-2 focus:ring-primary focus:border-primary outline-none transition-all duration-200"
          id="test_field"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-4"
      >
        <label
          class="block text-bodyText text-sm font-bold mb-2"
          for="test_dropdown"
        >
          Test Dropdown
        </label>
        <select
          class="border rounded w-full py-2 px-3 text-bodyText focus:ring-2 focus:ring-primary focus:border-primary outline-none transition-all duration-200 cursor-pointer"
          id="test_dropdown"
        >
          <option
            value=""
          >
            Select 
            Test Dropdown
          </option>
          <option
            value="Option 1"
          >
            Option 1
          </option>
          <option
            value="Option 2"
          >
            Option 2
          </option>
        </select>
      </div>
    </div>
    <button
      class="py-[7px] px-[30px] rounded-3xl transition-all duration-200 font-bold bg-primary text-white hover:bg-white hover:text-bodyText border-2 border-primary w-full bg-red-500 border-red-500 text-white hover:bg-white hover:text-red-500 cursor-not-allowed "
      disabled=""
      type="submit"
    >
      Please complete form
    </button>
  </form>
</div>
`;

exports[`Form matches snapshot with filled form data 1`] = `
<div>
  <form>
    <div
      class="mb-8"
    >
      <h2
        class="text-2xl font-semibold mb-4 text-bodyText"
      >
        Test Question
      </h2>
      <div
        class="mb-4"
      >
        <label
          class="block text-bodyText text-sm font-bold mb-2"
          for="test_field"
        >
          Test Field
        </label>
        <input
          class="appearance-none border rounded w-full py-2 px-3 text-bodyText focus:ring-2 focus:ring-primary focus:border-primary outline-none transition-all duration-200"
          id="test_field"
          type="text"
          value="test value"
        />
      </div>
      <div
        class="mb-4"
      >
        <label
          class="block text-bodyText text-sm font-bold mb-2"
          for="test_dropdown"
        >
          Test Dropdown
        </label>
        <select
          class="border rounded w-full py-2 px-3 text-bodyText focus:ring-2 focus:ring-primary focus:border-primary outline-none transition-all duration-200 cursor-pointer"
          id="test_dropdown"
        >
          <option
            value=""
          >
            Select 
            Test Dropdown
          </option>
          <option
            value="Option 1"
          >
            Option 1
          </option>
          <option
            value="Option 2"
          >
            Option 2
          </option>
        </select>
      </div>
    </div>
    <button
      class="py-[7px] px-[30px] rounded-3xl transition-all duration-200 font-bold bg-primary text-white hover:bg-white hover:text-bodyText border-2 border-primary w-full  "
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
